### PR TITLE
 Insert 'user-agent' header in requests done to the Elasticsearch server

### LINKED ISSUE #123: [Doc]Remove old compatibility notes
This note doesn't make sense in 7.x (or 6.x for that matter).  Documentation for older versions is still available if somebody needs it. 

Screen Shot 2020-04-17 at 11 07 34 AM

### PR DESCRIPTION
**Pull request by andsel**
## Release notes

Feat: add `user-agent` header passed to the Elasticsearch HTTP connection

## What does this PR do?

Pass an option to the underlying `Manticore` library containing the `user-agent` header. That header contains:
- Logstah version
- plugin version, asking to the `Gem` library
- OS name, version and CPU architechture
- JVM vendor name and version

Example:
```
logstash/7.14.1 (OS=Linux-5.4.0-84-generic-amd64; JVM=AdoptOpenJDK-11.0.11) logstash-input-elasticsearch/4.10.0
```

## Why is it important/What is the impact to the user?

no impact

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [ ] ~~I have made corresponding change to the default configuration files (and/or docker env variables)~~
- [x] I have added tests that prove my fix is effective or that my feature works

## Author's Checklist

- [ ]

## Related issues

- Relates https://github.com/elastic/logstash/issues/13206

### CODE REVIEW

File: lib/logstash/inputs/elasticsearch.rb [hunk 1]
```diff
@@ -407,6 +408,19 @@ def setup_api_key(api_key)
...
   end
 
+  def prepare_user_agent
+      os_name = java.lang.System.getProperty('os.name')
+      os_version = java.lang.System.getProperty('os.version')
+      os_arch = java.lang.System.getProperty('os.arch')
+      jvm_vendor = java.lang.System.getProperty('java.vendor')
+      jvm_version = java.lang.System.getProperty('java.version')
+
+#      plugin_version = Gem.loaded_specs["logstash-#{@plugin_type}-#{config_name}"].version
```
Comment (kaisecheng) [inline lib/logstash/inputs/elasticsearch.rb:418]: this comment can be cleaned

File: lib/logstash/inputs/elasticsearch/patches/_elasticsearch_transport_http_manticore.rb [hunk 2]
```diff
@@ -20,7 +20,18 @@ class Manticore
...
               headers = (options && options[:headers]) || {}
               headers[CONTENT_TYPE_STR] = find_value(headers, CONTENT_TYPE_REGEX) || DEFAULT_CONTENT_TYPE
-              headers[USER_AGENT_STR] = find_value(headers, USER_AGENT_REGEX) || user_agent_header
+
+              # this code is necessary to grab the correct user-agent header
+              # when this method is invoked with apply_headers(@request_options, options)
+              # from https://github.com/elastic/elasticsearch-ruby/blob/v7.14.0/elasticsearch-transport/lib/elasticsearch/transport/transport/http/manticore.rb#L113-L114
+              transport_user_agent = nil
+              if (options && options[:transport_options] && options[:transport_options][:headers])
+                transport_headers = {}
```
Comment (kaisecheng) [inline lib/logstash/inputs/elasticsearch/patches/_elasticsearch_transport_http_manticore.rb:29]: can delete this line? It got a new assignment in the next line.

**Review by kaisecheng — COMMENTED** [review]
I left two minor comments. LGTM for the user-agent and the test. Nice feature!

File: lib/logstash/inputs/elasticsearch/patches/_elasticsearch_transport_http_manticore.rb (hunk 2 above)
Comment (andsel) [reply lib/logstash/inputs/elasticsearch/patches/_elasticsearch_transport_http_manticore.rb:29]: good point!

**Review by kaisecheng — APPROVED** [review]

**Pull request merged**